# Phase 2

### Imports

In [1]:
from pyspark.sql import SparkSession
from pyspark.sql import functions as F

### Reading Data

In [2]:
spark = SparkSession.builder.getOrCreate()

hotel_booking = spark.read.csv(
    "../phase1_output_datasets/hotel_booking_cleaned.csv",
    header=True,
    inferSchema=True,
)
customer_reservations = spark.read.csv(
    "../phase1_output_datasets/customer_reservations.csv",
    header=True,
    inferSchema=True,
)
all_bookings = spark.read.csv(
    "../phase1_output_datasets/all_bookings.csv",
    header=True,
    inferSchema=True,
)

In [3]:
cols_to_check = [c for c in all_bookings.columns if c not in ("email", "booking_id")]

all_bookings = all_bookings.na.drop(subset=cols_to_check)

## 2.1 Spark Analysis

### Cancellation rates
Calculate cancellation rates for each month

The cancellation rate variable is like a boolean and the year and month columns are integers. To get the cancellation rate for each month we would first have to group all bookings by their year and month (year added because each year has the same 12 months). From there we would take the average of the cancellation rate column which gave the percentage of cancellations in the month.

In [ ]:
cancellation_rates = (all_bookings
    .groupBy("arrival_year", "arrival_month")
    .agg(F.avg("booking_status").alias("cancellation_rate"))
    .orderBy("arrival_year", "arrival_month"))

cancellation_rates.show()

+------------+-------------+-------------------+
|arrival_year|arrival_month|  cancellation_rate|
+------------+-------------+-------------------+
|        2015|            7| 0.4554670528602462|
|        2015|            8|0.41217436161980914|
|        2015|            9| 0.4087332808811959|
|        2015|           10| 0.3488513925594633|
|        2015|           11|0.20948275862068966|
|        2015|           12| 0.3355172413793103|
|        2016|            1| 0.2514672686230248|
|        2016|            2|0.34638319937775475|
|        2016|            3| 0.3082706766917293|
|        2016|            4| 0.3810140636565507|
|        2016|            5| 0.3503476033662642|
|        2016|            6|0.39678030303030304|
|        2016|            7|0.32798769771528996|
|        2016|            8| 0.3609022556390977|
|        2016|            9| 0.3756269738064276|
|        2016|           10| 0.4067357512953368|
|        2016|           11| 0.3686823104693141|
|        2016|      

### Averages
Compute average price and average number of nights for each month

This query was very similar to the previous one. We first started by grouping all bookings by month and year. From there we took the average of the "avg_price_per_room" column for each month.

The next query began similar with us grouping the bookings by month and year. But after that, there was an intermediate step where we added columns "stays_in_weekend_nights" and "stays_in_week_nights" together and put the results into a new column. We did this because the query wanted to calculate how many nights spent were spent in total for each booking. For there we took the average of the new column for each month group.

In [5]:
avg_month_price = (all_bookings
    .groupBy("arrival_year", "arrival_month")
    .agg(F.avg("avg_price_per_room").alias("avg_price_per_month"))
    .orderBy("arrival_year", "arrival_month"))

avg_month_price.show()

+------------+-------------+-------------------+
|arrival_year|arrival_month|avg_price_per_month|
+------------+-------------+-------------------+
|        2015|            7|  97.86348298334515|
|        2015|            8|  105.7779468661336|
|        2015|            9|  94.88720889063714|
|        2015|           10|  79.07416751372219|
|        2015|           11|  60.76284051724128|
|        2015|           12|  74.30403448275868|
|        2016|            1|  65.15584198645607|
|        2016|            2|   70.4027119522944|
|        2016|            3|  79.36027777777775|
|        2016|            4|  89.12324204293117|
|        2016|            5|  96.47828210757424|
|        2016|            6|  107.0391950757575|
|        2016|            7| 125.48952108963117|
|        2016|            8| 142.85367629600336|
|        2016|            9| 114.73706669143596|
|        2016|           10|  95.28716645077714|
|        2016|           11|  80.93624774368195|
|        2016|      

In [6]:
avg_nights = (all_bookings
    .withColumn(
        "total_nights",
        F.col("stays_in_weekend_nights") + F.col("stays_in_week_nights")
    )
    .groupBy("arrival_year", "arrival_month")
    .agg(F.avg("total_nights").alias("avg_nights_per_month"))
    .orderBy("arrival_year", "arrival_month"))

avg_nights.show()

+------------+-------------+--------------------+
|arrival_year|arrival_month|avg_nights_per_month|
+------------+-------------+--------------------+
|        2015|            7|  3.8664011585807385|
|        2015|            8|  3.6793912819190093|
|        2015|            9|   3.466365066876475|
|        2015|           10|   3.062614352510673|
|        2015|           11|              3.2875|
|        2015|           12|                3.05|
|        2016|            1|  2.7598194130925506|
|        2016|            2|  2.8885143894218306|
|        2016|            3|  3.2439431913116126|
|        2016|            4|  3.2041080680977054|
|        2016|            5|  3.2111233077204537|
|        2016|            6|   3.422727272727273|
|        2016|            7|   3.935413005272408|
|        2016|            8|  3.9066086268302334|
|        2016|            9|   3.457737321196359|
|        2016|           10|  3.1863665803108807|
|        2016|           11|   3.148691335740072|


### Monthly Bookings
Count monthly bookings by market segment. In categories, the term TA means Travel Agents and TO means Tour Operators

This query was interested in how many bookings were under each type in market_segment_type. We did this by grouping the data by market_segment_type then getting the count for each group.

In [7]:
# Online TA
# Offline TA/TO

bookings_by_market = (all_bookings
    .groupBy("market_segment_type")
    .agg(F.count("*").alias("num_bookings"))
    .orderBy(F.desc("num_bookings")))

bookings_by_market.show()

+-------------------+------------+
|market_segment_type|num_bookings|
+-------------------+------------+
|             Online|       56981|
|            Offline|       29028|
|             Groups|       13952|
|             Direct|        7848|
|          Corporate|        5593|
|      Complementary|         916|
|           Aviation|         252|
|          Undefined|           2|
+-------------------+------------+



### Seasonality
Identify the most popular month of the year for bookings based on revenue

This query wanted the most popular month based on revenue. I simply took the table from the query earlier (calculating avg_price_per_month for each month) and sorted it in descending order. The top row is the most popular month by revenue (August 2016) at average price $142.85.

In [8]:
avg_month_price.orderBy(F.desc("avg_price_per_month")).show()

+------------+-------------+-------------------+
|arrival_year|arrival_month|avg_price_per_month|
+------------+-------------+-------------------+
|        2016|            8| 142.85367629600336|
|        2016|            7| 125.48952108963117|
|        2018|            9| 122.90583051991898|
|        2018|            8| 121.21475884244347|
|        2018|            7| 115.44809151349233|
|        2016|            9| 114.73706669143596|
|        2018|            5| 113.81067744418755|
|        2018|            6| 111.96650015610335|
|        2018|           10|  111.4457226792006|
|        2016|            6|  107.0391950757575|
|        2015|            8|  105.7779468661336|
|        2017|            9| 103.33518496058223|
|        2018|            4| 101.99880116959012|
|        2015|            7|  97.86348298334515|
|        2018|           11|  96.53972138877005|
|        2016|            5|  96.47828210757424|
|        2016|           10|  95.28716645077714|
|        2018|      

In [9]:
avg_month_price.orderBy(F.desc("avg_price_per_month")).show(1)

+------------+-------------+-------------------+
|arrival_year|arrival_month|avg_price_per_month|
+------------+-------------+-------------------+
|        2016|            8| 142.85367629600336|
+------------+-------------+-------------------+
only showing top 1 row


# 2.2 Loading Data into PostgreSQL

First we start up the server in Docker

```shell
docker run -d --name bookings_db -e POSTGRES_USER=postgres -e POSTGRES_PASSWORD=password -e POSTGRES_DB=bookings -p 5432:5432 postgres
```

We then set up the tables in the PostgreSQL database by copying the SQL into the container and running it.

```shell
docker cp .\phase2\create_tables.sql bookings_db:/create_tables.sql
docker exec -it bookings_db psql -U postgres -d bookings -f /create_tables.sql
```

Then we can load the data from the cleaned CSV files we created using PySpark. This requires Hadoop; on Windows, the required setup can be done by downloading https://github.com/steveloughran/winutils and then setting the `HADOOP_HOME` environment variable to `path\to\winutils\hadoop-3.0.0\`. Then we use `spark-submit` to execute the script (support for running directly with `pyspark` was removed past pyspark 2).

```shell
spark-submit --packages org.postgresql:postgresql:42.7.8 .\phase2\spark_load_data.py
```

An easy way to check that our data was loaded successfully is to run Adminer, the PostgreSQL web UI.

```shell
docker run -p 8080:8080 adminer
```

Then we use the PostgreSQL information from our Docker command to log in to the database.

![Adminer login page](./adminer_login.png)

### Hotel Booking Schema

![Hotel booking schema](./hotel_booking_schema.png)

### Customer Reservations Schema

![Customer reservations schema](./customer_reservations_schema.png)

### All Bookings Schema

![All bookings schema](./all_bookings_schema.png)


# Data Loaded

We can check that all the data has been loaded correctly

![Hotel booking data](./hotel_booking_data.png)

![Customer reservations data](./customer_reservations_data.png)

![All bookings data](./all_bookings_data.png)